include/libtorrent/ssl/context.hpp

```diff
@@ -0,0 +1,279 @@
+{
+public:
+	using verify_context = base::verify_context;
+
+	explicit context(method m)
+		: base::context((static_cast<void>(global_init()), m))
+		, m_impl(std::make_shared<impl>(this)) {}
+
+	context(context&& other)
+		: base::context(std::move(other))
+		, m_impl(std::move(other.m_impl))
+	{
+		m_impl->parent = this;
+	}
+
+	context(const context& other) = delete;
+	~context() { if(m_impl) m_impl->parent = nullptr; }
+
+	context& operator=(context&& other)
+	{
+		base::context::operator=(std::move(other));
+		m_impl = std::move(other.m_impl);
+		m_impl->parent = this;
+		return *this;
+	}
+
+	error_code use_passphrase(std::string pass, error_code &ec)
+	{
+		set_password_callback([pass](std::size_t, password_purpose p)
+			{ return p == for_reading ? pass : ""; }
+			, ec);
+		return ec;
+	}
+
+	error_code set_verify_dn(std::string name, error_code& ec)
+	{
+		set_verify_callback(
+				std::bind(&context::verify_dn
+					, std::move(name)
+					, std::placeholders::_1
+					, std::placeholders::_2)
+				, ec);
+		return ec;
+	}
+
+	error_code set_trust_certificate(const const_buffer& certificate, file_format format, error_code &ec)
+	{
+#if defined TORRENT_USE_OPENSSL
+		if(format != pem)
+			return ec = {boost::asio::error::operation_not_supported, error::get_ssl_category()};
+
+		// create a new X.509 certificate store
+		X509_STORE* cert_store = X509_STORE_new();
+		if (!cert_store)
+			return ec = error_code(int(ERR_get_error()), error::get_ssl_category());
+
+		// wrap the PEM certificate in a BIO, for openssl to read
+		BIO* bp = BIO_new_mem_buf(
+			const_cast<void*>(static_cast<void const*>(certificate.data()))
+			, int(certificate.size()));
+
+		// parse the certificate into OpenSSL's internal representation
+		X509* cert = PEM_read_bio_X509_AUX(bp, nullptr, nullptr, nullptr);
+		BIO_free(bp);
+
+		if (!cert)
+		{
+			X509_STORE_free(cert_store);
+			return ec = error_code(int(ERR_get_error()), error::get_ssl_category());
+		}
+
+		// add cert to cert_store
+		X509_STORE_add_cert(cert_store, cert);
+		X509_free(cert);
+
+		// and lastly, replace the default cert store with ours
+		SSL_CTX_set_cert_store(native_handle(), cert_store);
+		return ec;
+#elif defined TORRENT_USE_GNUTLS
+		std::string cert(static_cast<char const*>(certificate.data()), certificate.size());
+
+        gnutls_datum_t ca;
+        ca.data = reinterpret_cast<unsigned char*>(const_cast<char*>(cert.c_str())); // must be null terminated
+        ca.size = cert.size();
+
+		// Warning: returns the number of certificates processed or a negative error code on error
+		int ret = gnutls_certificate_set_x509_trust_mem(native_handle()
+				, &ca
+				, format == pem ? GNUTLS_X509_FMT_PEM : GNUTLS_X509_FMT_DER);
+		return ec = ret >= 0 ? error_code() : error_code(ret, error::get_ssl_category());
+#endif
+	}
+
+#if defined TORRENT_USE_OPENSSL
+	error_code set_servername_callback(std::function<bool(stream_base& s, std::string name)> cb, error_code& ec)
```

Comment:
does this not have an implementation for GnuTLS?
Either way, this used to be a free function with the implementation hidden in a .cpp file, I think it's desirable to put everything that needs to include an openssl header in a .cpp file. I at least had attempted to do so previously, with the `openssl.cpp`

Reply:
This is already implemented in the GnuTLS wrapper itself (I put it there since for GnuTLS, SNI is not an extension API, it's in the core lib).
I'll put everything I can in .cpp files.

Reply:
Whether it's in the core API or not, the API *is* a callback still, right? I don't think it's important that callback itself is generic. Or rather, it's not clear to me the additional abstraction is worth making this callback SSL-library-agnostic.

Reply:
I think you are right, I'll reduce the abstraction by having two versions of the callback.